OrbitGl/ModuleData.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ModuleData.h"
+
+#include "FunctionUtils.h"
+#include "OrbitBase/Logging.h"
+#include "absl/synchronization/mutex.h"
+#include "capture_data.pb.h"
+
+using orbit_client_protos::FunctionInfo;
+
+FunctionInfo* ModuleData::FindFunctionByRelativeAddress(uint64_t relative_address,
+                                                        bool is_exact) const {
+  absl::MutexLock lock(&functions_mutex);
+  if (functions_.empty()) return nullptr;
+
+  uint64_t address = relative_address + load_bias();
+
+  if (is_exact) {
+    auto it = functions_.find(address);
+    return (it != functions_.end()) ? it->second.get() : nullptr;
+  }
+
+  auto it = functions_.upper_bound(address);
+  if (it == functions_.begin()) return nullptr;
+
+  --it;
+  FunctionInfo* function = it->second.get();
+  CHECK(function->address() <= address);
+
+  if (function->address() + function->size() < address) return nullptr;
+
+  return function;
+}
+
+void ModuleData::LoadSymbols(const orbit_grpc_protos::ModuleSymbols& module_symbols,
+                             uint64_t module_start_address) {
+  CHECK(!is_loaded());
+  absl::MutexLock lock(&functions_mutex);
+
+  uint32_t address_reuse_counter = 0;
+  for (const orbit_grpc_protos::SymbolInfo& symbol_info : module_symbols.symbol_infos()) {
+    auto [inserted_it, success] = functions_.try_emplace(
+        symbol_info.address(), FunctionUtils::CreateFunctionInfo(
+                                   symbol_info, load_bias(), file_path(), module_start_address));
+    FunctionInfo* function = inserted_it->second.get();
+    // It happens that the same address has multiple symbol names associated
+    // with it. For example: (all the same address)
+    // __cxxabiv1::__enum_type_info::~__enum_type_info()
+    // __cxxabiv1::__shim_type_info::~__shim_type_info()
+    // __cxxabiv1::__array_type_info::~__array_type_info()
+    // __cxxabiv1::__class_type_info::~__class_type_info()
+    // __cxxabiv1::__pbase_type_info::~__pbase_type_info()
+    if (success) {
+      hash_to_function_map_.emplace(FunctionUtils::GetHash(*function), function);
+    } else {
+      address_reuse_counter++;
+    }
+  }
+  if (address_reuse_counter != 0) {
+    LOG("Warning: %d absolute addresses are used by more than one symbol", address_reuse_counter);
```

Comment:
I think this can be smarter... for example, we can exclude symbols with 0 size (unless we already do that). Maybe there are other way we can get a better idea what is function and what is alias. which binary did you see it in?

Reply:
I actually printed the demangled names of these symbols before, but it is way too verbose. I don't know if we ignore 0 size elsewhere, out of the one that are found here, none of them are 0 size. 

In the Infiltrator demo this occurs 5419 times (total 357245 function symbols) (133 have size 1)
In other libraries it also happens: 
* 3155 duplicates out of 69291 symbols
* 225 out of 50661 
* 1658 out of 31393

I don't know why this happens. I just wanted to print something